src/client/chat.js

```diff
@@ -0,0 +1,538 @@
+        const toSign = []
+
+        for (let i = 0; i < 20; i++) {
+          const idx = (client.chat_log.tail + i) % 20
+          const entry = acknowledgements[idx]
+          if (entry) {
+            ackSet |= 1 << i
+            toSign.push(entry)
+            acknowledgements[idx] = { signature: entry.signature, pending: false }
+          }
+        }
+
+        const timestamp = BigInt(Date.now())
+        const salt = 0n
+        const bitset = Buffer.allocUnsafe(3)
+        bitset[0] = ackSet & 0xFF
+        bitset[1] = (ackSet >> 8) & 0xFF
+        bitset[2] = (ackSet >> 16) & 0xFF
+
+        client.write('chat_message', {
+          message,
+          timestamp,
+          salt,
+          signature: client.signMessage(message, timestamp, salt, toSign.map(ack => ack.signature)),
+          offset,
+          acknowledged: bitset
+        })
+      } else {
+        const timestamp = BigInt(Date.now())
+        client.write('chat_message', {
+          message,
+          timestamp,
+          salt: 0n,
+          signature: client.signMessage(message, timestamp)
+        })
+      }
+    } else client.write('chat', { message })
+  }
+
+  function fetchPreview (message) {
+    return new Promise((resolve, reject) => {
+      if (client.pendingPreview) {
+        if (options.queueMessages) queueMessagePreview(message, resolve, reject)
+        else reject(Error('Message could not be sent. Preview already in progress'))
+      } else {
+        requestPreview(message, resolve, reject)
+      }
+    })
+  }
+
+  function requestPreview (message, resolve, reject) {
+    client.previewId = client.previewId || 0
+    const id = client.previewId++
+    client.pendingPreview = { id, resolve, reject }
+
+    if (!client.lastPreviewRequest) client.lastPreviewRequest = 0
+    const diff = Date.now() - client.lastPreviewRequest
+
+    function sendPacket () {
+      client.write('chat_preview', {
+        query: id,
+        message
+      })
+
+      if (options.timeoutPreview) {
+        client.previewTimeout = setTimeout(() => {
+          throw Error('Server did not respond with preview in time') // Probably should not throw
+        }, options.timeoutPreview)
+      }
+    }
+
+    if (diff < 100) {
+      setTimeout(sendPacket, diff + 1)
+    } else sendPacket()
+  }
+
+  function queueMessagePreview (message, resolve, reject) {
+    if (!client.previewQueue) client.previewQueue = []
+    client.previewQueue.push({ message, resolve, reject })
+  }
+
+  // Packet handling
+
+  client.on('chat_preview', handleChatPreview)
+
+  client.on('player_info', handlePlayerInfo)
+
+  client.on('player_remove', handlePlayerRemove)
+
+  client.on('player_chat', handlePlayerChat)
+
+  client.on('system_chat', handleSystemChat)
+
+  client.on('hide_message', handleHideChat)
+
+  client.on('message_header', handleChatHeader)
+
+  client.on('should_display_chat_preview', packet => {
+    client.useChatPreview = packet.should_display_chat_preview
```

Comment:
this is not the same thing as useChatPreview, that's sent in server_data packet

Reply:
I see that you're right. The packets name simply indicates the behaviour that I assumed.

As I am reading it, the packet should actually be interpreted like this:

- If chat preview is currently enabled and should_display_chat_preview is false, chat preview should be disabled
- If chat preview is currently disabled and should_display_chat_preview is true, chat preview should be set to the value that was initially read from the server data packet

In the two other cases (useChatPreview === should_display_chat_preview), nothing changes